---
title: "LAVA"
author: "Gokberk Alagoz, L&G Dept., MPI for Psycholinguistics"
date: "`r Sys.Date()`"
output: rmarkdown::github_document
---

```{r setup, include=FALSE}
knitr::opts_chunk$set(echo = TRUE)
library(LAVA)
library(stringr)
library(kableExtra)
```

```{r processInput}
### Read in summary statistics and related info
input = process.input(input.info.file="/data/clusterfs/lag/users/gokala/beat-dyslexiaevol/results/LAVA/input/input.info.txt",           # input info file
                      sample.overlap.file="/data/clusterfs/lag/users/gokala/beat-dyslexiaevol/results/LAVA/input/sample.overlap.txt",   # sample overlap file (can be set to NULL if there is no overlap)
                      ref.prefix="/data/workspaces/lag/shared_spaces/Resource_DB/1KG_phase3/GRCh37/plink/1KG_phase3_GRCh37_EUR_allchr",                    # reference genotype data prefix
                      phenos=c("dyslexia","rhythm_impairment"))       # subset of phenotypes listed in the input info file that we want to process

# inspect the processed input data
ls(input)                     # this is actually an environment; hence ls() rather than str()
ls(input$sum.stats)   # processed summary statistics
head(input$sum.stats$bmi)
input$info            # processed input info file, with additional variables computed by process.input()
input$sample.overlap  # sample overlap file

head(input$ref$bim)   # bim file from reference data
  
# read more about this function
?process.input()
```

```{r prepareLocus}
# Prepare the loc file 
partitioned_genome = read.table("/data/workspaces/lag/workspaces/lg-genlang/Working/23andMe/Dyslexia2/Evolution/dys_rhy_pleiotropy/resources/LAVA_s2500_m25_f1_w200.blocks",
                                header = T)
colnames(partitioned_genome) = c("CHR", "START", "STOP")
partitioned_genome$LOC = 1:nrow(partitioned_genome)
partitioned_genome = partitioned_genome[,c(4,1,2,3)]
write.table(partitioned_genome,
            "/data/workspaces/lag/workspaces/lg-genlang/Working/23andMe/Dyslexia2/Evolution/dys_rhy_pleiotropy/resources/LAVA_s2500_m25_f1_w200.blocks.loci",
            row.names = F, quote = F)

### Read in locus info file
loci = read.loci("/data/workspaces/lag/workspaces/lg-genlang/Working/23andMe/Dyslexia2/Evolution/dys_rhy_pleiotropy/resources/LAVA_s2500_m25_f1_w200.blocks.loci")
head(loci)  # inspect the locus file
#>   LOC CHR     START      STOP
#> 1 100   1 113418038 114664387
#> 2 230   2  26894103  28819510
#> 3 266   2  57952946  59251996
#> 4 374   2 191051955 193033982
#> 5 464   3  47588462  50387742
#> 6 950   6  25684630  26396200
?read.loci  # read more about the function and possible data formats

### Create a locus object for the first locus to prepare it for analysis
locus = process.locus(loci[1,], input)

ls(locus)                               # inspect locus
c(locus$chr, locus$start, locus$stop)   # locus coordinates
str(locus$snps)                         # locus snps
locus$n.snps                            # N snps
locus$omega                             # genetic covariance matrix
locus$omega.cor                         # standardised genetic covariance matrix
locus$phenos                            # locus phenotypes

?process.locus  # find our more details in the function documentation
```

```{r univariateTest}
# for all phenotypes in the locus
run.univ(locus)

```

```{bash}
Rscript /data/workspaces/lag/workspaces/lg-genlang/Working/SCRIPTS/23andMe/clap-beat_dyslexia/pleiotropyevo/LAVA/lava_script.R "/data/workspaces/lag/shared_spaces/Resource_DB/1KG_phase3/GRCh37/plink/1KG_phase3_GRCh37_EUR_allchr" "/data/workspaces/lag/workspaces/lg-genlang/Working/23andMe/Dyslexia2/Evolution/dys_rhy_pleiotropy/resources/LAVA_s2500_m25_f1_w200.blocks.loci" "/data/clusterfs/lag/users/gokala/beat-dyslexiaevol/results/LAVA/input/input.info.txt" "/data/clusterfs/lag/users/gokala/beat-dyslexiaevol/results/LAVA/input/sample.overlap.txt" "dyslexia-rhythm_impairment" "dyslexia.rhythm_impairment"
```

```{r plotResults}
# 1) 2500 LD independent genomic bins
lava_res_s2500 = read.table("/data/clusterfs/lag/users/gokala/beat-dyslexiaevol/results/LAVA/output/LAVA_s2500_m25_f1_w200.blocks.bed.bivar.lava",
                          header = T)

lava_res_s2500$annot = "s2500"

# 2) neandertal depleted regions
lava_res_neanDep = read.table("/data/clusterfs/lag/users/gokala/beat-dyslexiaevol/results/LAVA/output/neanDepRegions_hg19.sorted.bed.bivar.lava",
                          header = T)

lava_res_neanDep$annot = "neanDep"

# 3) human-specific duplications
lava_res_humspedup = read.table("/data/clusterfs/lag/users/gokala/beat-dyslexiaevol/results/LAVA/output/human_specific_duplications_hg19-dennis_et_al.sorted.bed.bivar.lava",
                          header = T)

lava_res_humspedup$annot = "humspedup"

# merge and perform FDR correction
lava_merged = rbind(lava_res_s2500, lava_res_neanDep, lava_res_humspedup)
lava_merged$fdr = p.adjust(lava_merged$p, method = "fdr")

# save results in a table
write.table(lava_merged, "/data/clusterfs/lag/users/gokala/beat-dyslexiaevol/results/LAVA/output/LAVA_results_3annotations.txt", quote = F, row.names = F, sep = "\t")

```

